presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestSymbolMapper.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.sql.planner.optimizations;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.metadata.Metadata;
+import com.facebook.presto.spi.type.BigintType;
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.iterative.rule.test.PlanBuilder;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.TopNNode;
+import com.facebook.presto.sql.planner.plan.ValuesNode;
+import com.facebook.presto.testing.LocalQueryRunner;
+import com.google.common.collect.ImmutableList;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.testing.TestingSession.testSessionBuilder;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.airlift.testing.Closeables.closeAllRuntimeException;
+import static java.util.Collections.singleton;
+import static java.util.Collections.singletonList;
+import static org.testng.Assert.assertEquals;
+
+public class TestSymbolMapper
+{
+    private static final Symbol a = new Symbol("a");
+    private static final Symbol b = new Symbol("b");
+    private static final Symbol c = new Symbol("c");
+    private static final Symbol A = new Symbol("A");
+    private static final Symbol B = new Symbol("B");
+
+    private static final SymbolMapper SYMBOL_MAPPER = SymbolMapper.builder()
```

Comment:
You could make SymbolMapper handle cycles correctly (fail). And add test for that.
Probably a followup.